plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Vesting.hs

```diff
@@ -76,103 +89,93 @@ availableFrom (VestingTranche d v) range =
-        `PlutusTx.applyCode`
-            PlutusTx.liftCode v
-    where validatorParam vd = Scripts.wrapValidator (mkValidator vd)
+        remainingActual  = Validation.valueLockedBy ptx (Validation.ownHash ptx)
+
+        remainingExpected = 
+            remainingFrom vestingTranche1 pendingTxValidRange
+            + remainingFrom vestingTranche2 pendingTxValidRange
+
+    in remainingActual `Value.geq` remainingExpected
+            && Validation.txSignedBy ptx vestingOwner
+
+data Vesting
+instance Scripts.ScriptType Vesting where
+    type instance RedeemerType Vesting = ()
+    type instance DataType Vesting = ()
+
+vestingScript :: VestingParams -> ValidatorScript
+vestingScript = Scripts.validatorScript . scriptInstance
+
+scriptInstance :: VestingParams -> Scripts.ScriptInstance Vesting
+scriptInstance vesting = Scripts.Validator @Vesting
+    ($$(PlutusTx.compile [|| validate ||]) `PlutusTx.applyCode` PlutusTx.liftCode vesting)
+    $$(PlutusTx.compile [|| wrap ||])
+    where
+        wrap = Scripts.wrapValidator @() @()
+            
+contractAddress :: VestingParams -> Ledger.Address
+contractAddress = Scripts.scriptAddress . scriptInstance
+
+vestingContract :: VestingParams -> Contract VestingSchema T.Text ()
+vestingContract vesting = vest <|> retrieve where
+    retrieve = do
+        payment <- endpoint @"retrieve funds"
+        liveness <- retrieveFundsC vesting payment
+        case liveness of
+            Alive -> retrieve
+            Dead  -> pure ()
+    vest = endpoint @"vest funds" >> vestFundsC vesting
+
+payIntoContract :: VestingParams -> Value -> TxOut
+payIntoContract vp value = 
+    Tx.scriptTxOut' 
+        value 
+        (contractAddress vp)
+        (DataScript (PlutusTx.toData ()))
+
+vestFundsC
+    :: ( HasWriteTx s         
+       )
+    => VestingParams
+    -> Contract s T.Text ()
+vestFundsC vesting = do
+    let tx = unbalancedTx [] [payIntoContract vesting (totalAmount vesting)]
+    void $ writeTxSuccess tx
+
+data Liveness = Alive | Dead
+
+retrieveFundsC
+    :: ( HasAwaitSlot s
+       , HasUtxoAt s
+       , HasWriteTx s
+       )
+    => VestingParams
+    -> Value
+    -> Contract s T.Text Liveness
+retrieveFundsC vesting payment = do
+    let addr = contractAddress vesting
+    currentSlot <- awaitSlot 0
+    unspentOutputs <- utxoAt addr
+    let 
+        currentlyLocked = fold (AM.values unspentOutputs)
+        remainingValue = currentlyLocked - payment
+        liveness = if remainingValue `Value.gt` mempty then Alive else Dead
+        remainingOutputs = case liveness of
+                            Alive -> [payIntoContract vesting remainingValue]
+                            Dead  -> []
+        tx = Typed.collectFromScriptFilter (\_ _ -> True) unspentOutputs (scriptInstance vesting) ()
+                & validityRange .~ Interval.from currentSlot
+                & requiredSignatures .~ [vestingOwner vesting]
+                & outputs .~ remainingOutputs
```

Comment:
I confused myself for a bit trying to work out why we weren't paying back to ourselves, and then I realised balancing does that. IMO it would be better to be explicit, but worth a comment at least.